### PR TITLE
Short-cut rules where possible & relax reasoner planner orderings

### PR DESCRIPTION
**Pull request by krishnangovindraj**
## What is the goal of this PR?
We revise the criteria for a constraint to be dependent on a variable being bound. This refines the search-space of the reasoner planner to include more legal plans, and allows an optimisation for reasoning involving relations playing roles in other relations.

In certain cases during reasoning, we can determine that all answers for a given constraint have already been computed. In such cases, we can bypass inference and simply look up the answers.

## What are the changes implemented in this PR?
* Loosen the criteria for resolvable dependencies:
  * A resolvable _r_ is only dependent on a variable _v_ if : 
      1. _v_ is 'generated' by some other concludable in the conjunction AND
      2. _r_ contains a constraint involving `v`
      3. no constraint in _r_  generates values for `v` 
      * (Currently, predicates are the only constraints which can involve a variable but not generate inferred values for it)
  * Negations remain dependent on variables which occur unnegated in conjunction(s) which contain it.    
* `RecursivePlanner` no longer relaxes dependencies to generate orderings when it cannot schedule any resolvable.
* `ConcludableProcessor.Match` will not trigger rules (i.e. create conclusion processors upstream)  if the answer is already guaranteed to be materialised. It will create instead do a traversal which includes inferred concepts.

### CODE REVIEW

File: reasoner/controller/ConcludableController.java [hunk 1]
```diff
@@ -180,20 +180,22 @@ protected Driver<ConclusionController.Explain> registerConclusionController(Rule
...
+        final ConceptMap bounds;
         private final Set<Variable.Retrievable> unboundVars;  // TODO: Can just use a boolean to indicate if fully bound
         private final Map<Conclusion, Set<Unifier>> conclusionUnifiers;
         private final Set<Identifier> requestedConnections;
         final java.util.function.Supplier<FunctionalIterator<ConceptMap>> traversalSuppplier;
 
         Processor(Driver<PROCESSOR> driver,
                   Driver<? extends AbstractController<?, INPUT, OUTPUT, REQ, PROCESSOR, ?>> controller,
-                  Context context, ConceptMap bounds,
+                  Concludable concludable, Context context, ConceptMap bounds,
```
Comment (flyingsilverfin) [inline reasoner/controller/ConcludableController.java:192]: Concludable should come after context

File: reasoner/controller/ConcludableController.java [hunk 2]
```diff
@@ -241,22 +243,33 @@ private void mayRequestConnection(REQ conclusionRequest) {
...
+
+
+            @Override
+            public void setUp() {
+                // TODO: Add a find first optimisation when all variables are bound
+                if (concludable.isRelation() && concludable.generating().isPresent() && bounds.contains(concludable.generating().get().id())) {
+                    // Optimisation: If we know the relation instance, just do a lookup.
+                    setHubReactive(fanInFanOut(this));
+                    mayAddTraversal(true);
+                } else {
```
Comment (flyingsilverfin) [inline reasoner/controller/ConcludableController.java:264]: we can generalise this optimisation:
```
                if (concludable.isInferredAnswer(bounds)) {
                    // don't trigger rules ever
                    if (!bounds are fully binding concludable) {
                        mayAddTraversal(true);
                    } else {
                        return singleton bounds as answer
                    }
                } else {
                    mayAddTraversal(false);
                    triggerRules();
                }
```

you'll need to add `Concludable.isAnswer` which returns true if the concept map provided is an answer that fits as either an inferred or non-inferred answer

File: reasoner/planner/ReasonerPlanner.java [hunk 3]
```diff
@@ -95,38 +101,47 @@ Plan getPlan(CallMode callMode) {
 
     abstract Plan computePlan(CallMode callMode);
 
+    private static boolean isDependency(Resolvable<?> resolvable, Variable v) {
```
Comment (flyingsilverfin) [inline reasoner/planner/ReasonerPlanner.java:104]: put below the user

File: reasoner/controller/ConcludableController.java (hunk 1 above)
Comment (krishnangovindraj) [reply reasoner/controller/ConcludableController.java:191]: It comes before in the controller constructor. Should I change both?

File: reasoner/planner/ReasonerPlanner.java [hunk 4]
```diff
@@ -95,38 +101,47 @@ Plan getPlan(CallMode callMode) {
...
+
+        // Add dependency between negateds and shared variables
+        Set<Variable> unnegatedVars = iterate(resolvables).filter(r -> !r.isNegated()).flatMap(r -> iterate(r.variables())).toSet();
+        iterate(resolvables).filter(Resolvable::isNegated)
+                .forEachRemaining(resolvable -> deps.get(resolvable).addAll(intersection(resolvable.variables(), unnegatedVars)));
+
+        // Add dependency for resolvable to any variables for which it can't (independently) generate all satisfying values
+        iterate(resolvables).filter(resolvable -> !resolvable.isNegated())
+                .forEachRemaining(resolvable -> deps.get(resolvable).addAll(
+                        iterate(resolvable.variables()).filter(ReasonerPlanner::hasNoGeneratingConstraint).toSet()
```
[the remark this answers is not in the document]
Comment (krishnangovindraj) [reply reasoner/planner/ReasonerPlanner.java:120]: The RHS part doesn't seem to hold. 

The BDD scenario "a rule can be applied based on concept inequality" has a query
```
        $value isa! retailer;
        $unwantedValue isa! retailer;
        $x has $value;
        $y has $unwantedValue;
        $unwantedValue "Ocado";
        $value != $unwantedValue;
```
which is (strangely) planned to be:
1. `Concludable: $y has $unwantedValue;`
2. `Retrievable: $value != $unwantedValue;`
3. `Concludable: $x has $value;` 

2 being a separate retrievable is part of the problem. Another problem is a weak planning of predicates and the connectedness approach to planning.

File: reasoner/planner/RecursivePlanner.java [hunk 5]
```diff
@@ -373,10 +374,6 @@ private void porDFS(Stack<Resolvable<?>> currentPath, Set<Variable> currentBound
                     .filter(r -> ReasonerPlanner.dependenciesSatisfied(r, currentBoundVars, dependencies) && !sleepSet.contains(r))
                     .toList();
 
-            if (enabled.isEmpty()) {
-                enabled = iterate(remaining).filter(r -> !r.isNegated() && !sleepSet.contains(r)).toList();
```
Comment (krishnangovindraj) [inline reasoner/planner/RecursivePlanner.java:377]: When our dependencies were too strict, we'd relax them to be able to generate orderings.
Now that they're strict enough, We should prevent that.

File: reasoner/controller/ConcludableController.java [hunk 6]
```diff
@@ -241,22 +234,44 @@ private void mayRequestConnection(REQ conclusionRequest) {
...
+                    addRules();
+                }
+            }
+
+            private boolean canBypassReasoning(ConceptMap bounds) {
+                if (concludable.isHas()) {
+                    Concept owner; Concept attribute;
+                    return  (owner = bounds.get(concludable.asHas().owner().id())) != null &&
+                            (attribute = bounds.get(concludable.asHas().attribute().id())) != null &&
+                            (owner.asThing().hasInferred(attribute.asAttribute()) || owner.asThing().hasNonInferred(attribute.asAttribute()));
```
Comment (krishnangovindraj) [inline reasoner/controller/ConcludableController.java:263]: We check the existence of the edge here, to determine whether we can bypass traversal.
Will see how hard it is to bypass the full traversal for non-relation cases.

**Review by flyingsilverfin — APPROVED** [review]

**Pull request merged**